utils/ixdtf/src/parsers/mod.rs

```diff
@@ -0,0 +1,241 @@
+
+//! The parser module contains the implementation details for `IXDTFParser` and `ISODurationParser`
+
+use crate::{ParserError, ParserResult};
+
+extern crate alloc;
+
+#[cfg(feature = "duration")]
+use records::DurationParseRecord;
+use records::IxdtfParseRecord;
+
+pub mod records;
+
+mod annotations;
+pub(crate) mod datetime;
+#[cfg(feature = "duration")]
+pub(crate) mod duration;
+mod grammar;
+mod time;
+pub(crate) mod timezone;
+
+#[cfg(test)]
+mod tests;
+
+/// `assert_syntax!` is a parser specific utility macro for asserting a syntax test, and returning the
+/// the provided provided error if the assertion fails.
+#[macro_export]
+macro_rules! assert_syntax {
+    ($cond:expr, $err:ident $(,)?) => {
+        if !$cond {
+            return Err(ParserError::$err);
+        }
+    };
+}
+
+/// Parsing options that can be provided to `IxdtfParser` to change parsing behavior.
+#[non_exhaustive]
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum IxdtfOptions {
+    /// No modified parsing, i.e. parses as a date time
+    None,
+    /// Parses the value as an annotated YearMonth.
+    YearMonth,
+    /// Parses the value as an annotated MonthDay.
+    MonthDay,
+    /// Parses the value as an annotated Time.
+    Time,
+}
+
+/// `IxdtfParser` is the primary parser implementation of `ixdtf`.
+///
+/// This parser provides various options for parsing date/time strings with the extended notation
+/// laid out in [sedate's IXDTF][ixdtf-draft] along with other variations laid out in the [`Temporal`][temporal-proposal].
+///
+/// [ixdtf-draft]: https://datatracker.ietf.org/doc/draft-ietf-sedate-datetime-extended/
+/// [temporal-proposal]: https://tc39.es/proposal-temporal/
+#[derive(Debug)]
+pub struct IxdtfParser<'a> {
+    options: IxdtfOptions,
+    cursor: Cursor<'a>,
+}
+
+impl<'a> IxdtfParser<'a> {
+    /// Creates a new `IXDTFParser` from a provided `&str`.
+    pub fn new(value: &'a str) -> Self {
+        Self {
+            options: IxdtfOptions::None,
+            cursor: Cursor::new(value),
+        }
+    }
+
+    /// Set an additional parser option.
+    pub fn with_option(mut self, option: IxdtfOptions) -> Self {
+        self.options = option;
+        self
+    }
+
+    /// Parses the source as a [DateTime string][temporal-dt].
+    ///
+    /// This is the baseline parser where the TimeRecord, UTCOffset, and all annotations are optional.
+    ///
+    /// [temporal-dt]: https://tc39.es/proposal-temporal/#prod-TemporalDateTimeString
+    pub fn parse(&mut self) -> ParserResult<IxdtfParseRecord<'a>> {
+        match self.options {
+            IxdtfOptions::None => datetime::parse_annotated_date_time(&mut self.cursor),
+            IxdtfOptions::YearMonth => datetime::parse_annotated_year_month(&mut self.cursor),
+            IxdtfOptions::MonthDay => datetime::parse_annotated_month_day(&mut self.cursor),
+            IxdtfOptions::Time => time::parse_annotated_time_record(&mut self.cursor),
+        }
```

Comment:
Thought: With big match statements like this, it is better for DCE if we instead had four functions like `parse`, `parse_year_month`, `parse_month_day`, and `parse_time`. Then we wouldn't need the enum.